### ISSUE TITLE
White line issue on pdf file content.

### ISSUE THREAD

**Issue by noor-ul-ain11** [mozilla/pdf.js#18558]
I'm rendering pdf file content on the canvas and facing some white lines on the content as shown in the following image.
![Screenshot (200)](https://github.com/user-attachments/assets/c05155b2-8337-47b8-b8cc-6de1268aea3b)

here is the code I'm using. 
<?php
require_once get_stylesheet_directory() . '/vendor/setasign/fpdf/fpdf.php';
require_once get_stylesheet_directory() . '/vendor/setasign/fpdi/src/autoload.php';
require_once ABSPATH . 'wp-load.php';

use setasign\Fpdi\Fpdi;

function get_user_order_ids($current_user_id) {
    global $wpdb;
    $results = $wpdb->get_col($wpdb->prepare(
        "SELECT id 
         FROM {$wpdb->prefix}wc_orders 
         WHERE customer_id = %d 
         AND status = 'wc-active'",
        $current_user_id
    ));
    return $results;
}

function user_has_order($sub_id, $current_user_id) {
    $order_ids = get_user_order_ids($current_user_id);
    return in_array($sub_id, $order_ids);
}

function download_pdf($url, $save_path) {
    $ch = curl_init($url);
    $fp = fopen($save_path, 'wb');
    curl_setopt($ch, CURLOPT_FILE, $fp);
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
    curl_exec($ch);
    curl_close($ch);
    fclose($fp);
}

function append_text_to_pdf($file_path, $metadata) {
    $pdf = new Fpdi();
    $pageCount = $pdf->setSourceFile($file_path);

    for ($pageNo = 1; $pageNo <= $pageCount; $pageNo++) {
        $templateId = $pdf->importPage($pageNo);
        $size = $pdf->getTemplateSize($templateId);

        $pdf->AddPage($size['orientation'], [$size['width'], $size['height']]);
        $pdf->SetMargins(0, 0, 0);
        $pdf->SetAutoPageBreak(false, 0);
        $pdf->useTemplate($templateId);

        $pdf->SetFont('Arial', '', 9);

        $pdf->SetXY(50, $size['height'] - 15);
        $pdf->Cell(0, 10, $metadata['student_name'], 0, 0, 'L');

        $pdf->SetXY(($size['width'] / 2) - 100, $size['height'] - 15);
        $pdf->Cell(0, 10, $metadata['student_school_name'], 0, 0, 'C');

        $pdf->SetXY($size['width'] - 90, $size['height'] - 15);
        $pdf->Cell(0, 10, $metadata['schedule_dates'], 0, 0, 'C');
        
        $copyright_text = '© D. Cohen. All Rights Reserved. Copying or distribution prohibited without the express prior permission of D. Cohen.';
        $offset = 20;
        $pdf->SetXY((($size['width'] - $pdf->GetStringWidth($copyright_text)) / 2) - $offset, $size['height'] - 10);
        $pdf->Cell(0, 10, utf8_decode($copyright_text), 0, 0, 'C');
        
    }

    return $pdf->Output('S');
}

function format_schedule_dates($start_date, $next_payment_date) {
    if (!$start_date || $start_date === '0') {
        $start_formatted = 'N/A';
    } else {
        $start_dt = new DateTime($start_date);
        $start_formatted = $start_dt->format('n/Y');
    }

    if (!$next_payment_date || $next_payment_date === '0') {
        $next_payment_formatted = 'N/A';
    } else {
        $next_payment_dt = new DateTime($next_payment_date);
        $next_payment_formatted = $next_payment_dt->format('n/Y');
    }

    return "$start_formatted-$next_payment_formatted";
}

$current_user_id = get_current_user_id();

if (isset($_GET['file']) && !empty($_GET['file']) && isset($_GET['sub_id']) && !empty($_GET['sub_id'])) {
    $pdf_url = urldecode(sanitize_text_field($_GET['file']));
    $local_pdf_path = '/tmp/downloaded.pdf';
    download_pdf($pdf_url, $local_pdf_path);

    $sub_id = intval($_GET['sub_id']);
    if (!user_has_order($sub_id, $current_user_id)) {
        wp_die('You do not have access to this file.');
    }
    global $wpdb;
    $order_id = intval($_GET['order_id']);

    $student_first_name = $wpdb->get_var($wpdb->prepare("SELECT first_name FROM {$wpdb->prefix}student_info WHERE order_id = %d ", $order_id));
    $student_last_name = $wpdb->get_var($wpdb->prepare("SELECT last_name FROM {$wpdb->prefix}student_info WHERE order_id = %d", $order_id));
    $student_school_name = $wpdb->get_var($wpdb->prepare("SELECT school_name FROM {$wpdb->prefix}student_info WHERE order_id = %d ", $order_id));
    $schedule_start = $wpdb->get_var($wpdb->prepare("SELECT meta_value FROM {$wpdb->prefix}wc_orders_meta WHERE order_id = %d AND meta_key = '_schedule_start'", $sub_id));
    $schedule_next_payment = $wpdb->get_var($wpdb->prepare("SELECT meta_value FROM {$wpdb->prefix}wc_orders_meta WHERE order_id = %d AND meta_key = '_schedule_end'", $sub_id));

    $schedule_dates = format_schedule_dates($schedule_start, $schedule_next_payment);

    $metadata = [
        'student_name' => trim($student_first_name . ' ' . $student_last_name),
        'student_school_name' => $student_school_name,
        'schedule_dates' => $schedule_dates,
    ];

    $modified_pdf_content = append_text_to_pdf($local_pdf_path, $metadata);
    $modified_pdf_base64 = base64_encode($modified_pdf_content);
?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Custom PDF Viewer</title>
    <style>
        @media print {
            body {
                padding: 0;
                margin: 0;
            }
            .header, .footer, .sidebar {
                display: none;
            }
            .page {
                page-break-before: always;
            }
            @page {
                margin: 0;
            }
            html, body {
                margin: 0;
                padding: 0;
                height: 100%;
            }
        }
        html, body {
            margin: 0;
            padding: 0;
            height: 100%;
            overflow: hidden;
        }
        #viewer {
            width: 100%;
            height: 100%;
            overflow-y: scroll;
            background-color: #f0f0f0;
            display: flex;
            flex-direction: column;
            align-items: center;
            gap: 20px;
            padding: 20px;
            box-sizing: border-box;
            opacity: 99.9%;
        }
        .page-container {
            position: relative;
            box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);

        }
        .print-button {
            position: absolute;
            top: 20px;
            right: 20px;
            padding: 10px 20px;
            background-color: #007bff;
            color: white;
            border: none;
            border-radius: 5px;
            cursor: pointer;
            z-index: 100;
        }
        .page {
            margin: 0;
            padding: 0;
            border: none;
            box-sizing: border-box;
        }
    </style>
</head>
<body>
    <div id="viewer"></div>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/pdf.js/2.14.305/pdf.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js"></script>
    <script>
        // Base64 encoded PDF data
        var pdfData = atob('<?php echo $modified_pdf_base64; ?>');

        // Load the PDF data
        pdfjsLib.getDocument({ data: pdfData }).promise.then(function(pdf) {
            var viewer = document.getElementById('viewer');
            var renderPromises = [];

            // Function to render a page
            function renderPage(page, pageIndex) {
                var scale = 1.5; // Adjust as needed
                var viewport = page.getViewport({ scale: scale });

                var headerHeight = 0; // Adjust as per your PDF's header height
                var footerHeight = 0; // Adjust as per your PDF's footer height

                // Adjusted viewport without header and footer
                var adjustedViewport = page.getViewport({ scale: scale, offsetY: headerHeight, offsetBottom: footerHeight });

                // Create container for the page and print button
                var pageContainer = document.createElement('div');
                pageContainer.className = 'page-container';

                // Create canvas element
                var canvas = document.createElement('canvas');
                canvas.className = 'page';
                pageContainer.appendChild(canvas);

                var context = canvas.getContext('2d');
                canvas.height = adjustedViewport.height;
                canvas.width = adjustedViewport.width;

                // Render PDF page into canvas context
                var renderContext = {
                    canvasContext: context,
                    viewport: adjustedViewport,
                    transform: [1, 0, 0, 1, 0, -headerHeight] // Translate the rendering context to remove header
                };

                // Append page container to viewer
                viewer.appendChild(pageContainer);

                // Render PDF page
                return page.render(renderContext).promise.then(function() {
                    // Add print button to page container
                    var printButton = document.createElement('button');
                    printButton.className = 'print-button';
                    printButton.id = 'pdf-print-button';
                    printButton.textContent = 'Print';
                    printButton.addEventListener('click', function() {
                        printPage(canvas);
                    });
                    pageContainer.appendChild(printButton);
                });
            }

            // Queue render tasks for all pages
            for (var pageNum = 1; pageNum <= pdf.numPages; pageNum++) {
                renderPromises.push(pdf.getPage(pageNum).then(function(page) {
                    return renderPage(page, pageNum - 1);
                }));
            }

            // Ensure all pages are rendered
            Promise.all(renderPromises).then(function() {
                console.log('All pages rendered successfully.');
            }).catch(function(error) {
                console.error('Error occurred while rendering pages:', error);
            });
        }).catch(function(error) {
            console.error('Error loading PDF:', error);
        });

        // Print specific page function
        function printPage(canvas) {
            // Create a new window for printing
            var printWindow = window.open('', '_blank');

            // Ensure printWindow is created successfully
            if (!printWindow) {
                console.error('Failed to open print window.');
                return;
            }

            // Write print content to the new window
            var printDocument = printWindow.document;
            printDocument.open();
            printDocument.write('<html><head><title>Print PDF Page</title>');
            printDocument.write('<style>@page { margin: 0; }</style>'); // Remove margins
            printDocument.write('</head><body style="margin: 0; padding: 0;">');

            // Create an image element to hold the canvas data
            var img = new Image();
            img.src = canvas.toDataURL();
            img.style.width = '100%';

            // Ensure img element is loaded before writing it to the document
            img.onload = function() {
                printDocument.body.appendChild(img);
                printDocument.write('</body></html>');
                printDocument.close();

                // Delay print invocation to ensure content is fully written
                setTimeout(function() {
                    printWindow.print();
                    printWindow.close();
                }, 1000); // Adjust delay as needed
            };

            img.onerror = function() {
                console.error('Failed to load image for printing.');
            };

            // Append image element to the document
            printDocument.body.appendChild(img);
        }

        document.getElementById('viewer').addEventListener('click', function(event) {
            if (event.target.classList.contains('print-button')) {
                console.log('Button clicked:', event.target.id);
                var canvas = event.target.closest('.page-container').querySelector('canvas');
                printPage(canvas);

                // Retrieve order_id and stu_id from the URL
                const urlParams = new URLSearchParams(window.location.search);
                const orderId = urlParams.get('order_id');
                const stuId = urlParams.get('stu_id');

                // Send AJAX request to increment the printable_clicks
                if (orderId && stuId) {
                    jQuery.ajax({
                        url: '<?php echo admin_url('admin-ajax.php'); ?>',
                        type: 'GET',
                        data: {
                            action: 'increment_clicks_pdf',
                            order_id: orderId,
                            stu_id: stuId,
                            column: 'printable_clicks',
                            security: '<?php echo wp_create_nonce('increment_clicks_nonce'); ?>'
                        },
                        success: function(response) {
                            if (response.success) {
                                console.log('Click count incremented successfully.');
                            } else {
                                console.log('Error incrementing click count:', response.data);
                            }
                        },
                        error: function(xhr, status, error) {
                            console.error('AJAX request failed:', status, error);
                        }
                    });
                } else {
                    console.error('Invalid order ID or student ID.');
                }
            }
        });
    </script>
</body>
</html>

<?php
    exit;
}
?>   Kindly help me resolve this issue.
